Add a new extension: VK_EXT_map_memory_placed (#1906)
KhronosGroup/Vulkan-Docs: chapters/memory.txt

  * [[VUID-vkMapMemory-memory-00683]]
    pname:memory must: not have been allocated with multiple instances
endif::VK_KHR_device_group[]
ifdef::VK_EXT_map_memory_placed[]
  * ename:VK_MEMORY_MAP_PLACED_BIT_EXT must: not be set in pname:flags
endif::VK_EXT_map_memory_placed[]
****

include::{generated}/validity/protos/vkMapMemory.txt[]
--

ifdef::VK_EXT_map_memory_placed[]
[open,refpage='VkMemoryMapFlagBits',desc='Bitmask specifying additional parameters of a memory map',type='enums']
--

> Bits which can: be set in flink:vkMapMemory::pname:flags, specifying

Reviewer: Feels a bit weird to call out vkMapMemory when the only flag explicitly _cannot_ be used with it, maybe:

```suggestion
Bits which can: be set in flink:vkMapMemory::pname:flags and flink:vkMapMemory2EXT::pname:flags, specifying
```

Reviewer 2: Done.

Reviewer 3: N.b. it's helpful to have back-references to all the users of a type, but it's fragile against extensions, and can get wordy. In the refpages it automatically generates "See Also" references for all APIs which use the API being described but we don't do that in the spec itself (though we could).

    memory object.
  * pname:size is the size of the memory range to map, or
    ename:VK_WHOLE_SIZE to map from pname:offset to the end of the
    allocation.
  * pname:flags is reserved for future use.

.Valid Usage
****
  * pname:memory must: not be currently host mapped
  * pname:offset must: be less than the size of pname:memory
  * If pname:size is not equal to ename:VK_WHOLE_SIZE, pname:size must: be
    greater than `0`
  * If pname:size is not equal to ename:VK_WHOLE_SIZE, pname:size must: be
    less than or equal to the size of the pname:memory minus pname:offset
  * pname:memory must: have been created with a memory type that reports
    ename:VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT
ifdef::VK_KHR_device_group[]
  * pname:memory must: not have been allocated with multiple instances
endif::VK_KHR_device_group[]
ifdef::VK_EXT_map_memory_placed[]
  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags, the
    <<features-memoryMapPlaced, pname:memoryMapPlaced>> feature must: be
    enabled
  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags, the
    pname:pNext chain must: include a slink:VkMemoryMapPlacedInfoEXT
    structure

>   * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags and the

Reviewer 4: What's the point of gating this functionality?  What would prevent any implementation style from subtracting the offset from the desired pointer, and mapping the object starting from the beginning?

Reviewer 5: Then it would use more address space than the application expects.

Reviewer 2: Yeah, if you're going to be mapping over an explicit range, the map has to match exactly.  Otherwise you'll stomp things.  And, sadly, most Linux kernel drivers don't support an `mmap()` offset today.  I'd like to see that fixed but, even if I typed the kernel patches today, it would be a year or two before they hit users so we need the feature bit.

Reviewer 4: Okay, and another question, should it be documented how much address space preceding the mapping is used up by vkMapMemory, so we know where and how big to make the reserved mapping?  Would the mapping always just start in the page of the address passed in? If not, there could be a problem w/ MAP_FIXED overwriting preceding pages unexpectedly.

Reviewer 2: The current plan, if we can get away with it, is that it will map exactly the requested range, like `mmap()` does.  The usual mapping system in DRM allows you to map as little or as much as you want at the beginning of the BO, you just can't offset into it.  As long as any extra space the driver needs is placed at the end, it shouldn't be a problem.

If we need to add something like that, we can, but I'd like to avoid it if we can.

Reviewer 4: Ah I see, and I just noticed the offset alignment requirement a few lines down, perfect.

  * [[VUID-vkMapMemory-memory-00683]]
    pname:memory must: not have been allocated with multiple instances
endif::VK_KHR_device_group[]
ifdef::VK_EXT_map_memory_placed[]
  * ename:VK_MEMORY_MAP_PLACED_BIT_EXT must: not be set in pname:flags
endif::VK_EXT_map_memory_placed[]
****

include::{generated}/validity/protos/vkMapMemory.txt[]
--

ifdef::VK_EXT_map_memory_placed[]
[open,refpage='VkMemoryMapFlagBits',desc='Bitmask specifying additional parameters of a memory map',type='enums']
--
Bits which can: be set in flink:vkMapMemory::pname:flags and
slink:VkMemoryMapInfoEXT::pname:flags, specifying additional properties of
a memory map, are:

include::{generated}/api/enums/VkMemoryMapFlagBits.txt[]

  * ename:VK_MEMORY_MAP_PLACED_BIT_EXT requests that the implementation
    attempt to place the memory map at the virtual address specified by the

>     client via pname:ppData.

Reviewer: ```suggestion
    client via pname:ppData.
    Can: only be used in slink:VkMemoryMapInfoEXT::pname:flags.
```

Worth calling this out?

Reviewer 2: We can, I guess.  There's a VU for that but we can call it out explicitly, I suppose.  IDK.

Reviewer: I think a naive reader might get confused but probably not for long so eh, take it or leave it.

    ename:VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_ALLOCATION_BIT_EXT or
    ename:VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_MAPPED_FOREIGN_MEMORY_BIT_EXT
endif::VK_EXT_external_memory_host[]
endif::VK_EXT_map_memory_placed[]
****

include::{generated}/validity/structs/VkMemoryMapInfoEXT.txt[]
--
endif::VK_EXT_map_memory2[]

ifdef::VK_EXT_map_memory_placed[]
[open,refpage='VkMemoryMapPlacedInfoEXT',desc='Structure containing memory map placement parameters',type='structs']
--
If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in
sname:VkMemoryMapInfoEXT::pname:flags and the pname:pNext chain of
slink:VkMemoryMapInfoEXT includes a sname:VkMemoryMapPlacedInfoEXT
structure, then that structure specifies the placement address of the
memory map.
The implementation will attempt to place the memory map at
that address.
If successful, pname:ppData will be set to the same value as
sname:VkMemoryMapPlacedInfoEXT::pname:pPlacedAddress and
fname:vkMapMemory2EXT will return ename:VK_SUCCESS.
If it cannot place the map at the requested address for any
reason, the memory object is left unmapped and fname:vkMapMemory2EXT will
return ename:VK_ERROR_MEMORY_MAP_FAILED.

[NOTE]
.Note
====

> In order for a placed map to succeed, the client may: first have to reserve

Reviewer: ```suggestion
In order for a placed map to succeed, the client may first have to reserve
```

We don't use normative tags in notes.

Reviewer: Although thinking about it - should this be a note? Also is this something we should advertise via a limit if it's a static property, or is this more "if you haven't reserved it, then it might have already been snatched up by another allocation"?

Reviewer 2: Hrm... Looking at the `mmap()` docs, I suspect we may may actually want two flags:

 - `VK_MEMORY_MAP_TRY_PLACED`: Attempts to place the memory map and fails if another mapping currently exists in that range.
 - `VK_MEMORY_MAP_PLACED_REPLACE`: Replaces whatever is at the current location regardless.  This needs a VU that it's not overwritting another Vulkan mapping because we want the "one map at a time" rule to be validatable.

These map to `MAP_FIXED_NOREPLACE` and `MAP_FIXED`, respectively.  Then it would be explicit what happens and I think we could drop this note entirely.  We could also make them different feature bits.

Reviewer 5: It's simple for applications to emulate `VK_MEMORY_MAP_TRY_PLACED` with a `mmap` call with `MAP_FIXED_NOREPLACE` + `VK_MEMORY_MAP_PLACED_REPLACE`. So I see no reason to have `VK_MEMORY_MAP_TRY_PLACED` at all.

Reviewer 2: That's not going to be atomic.  If something else successfully maps between the two calls, the `PLACED_REPLACE` will stomp it.  I don't know if that's a problem or not.

Reviewer 4: Whether a flag is added to Vulkan or not, a component using MMAP_FIXED/PLACE_REPLACE on an address it hasn't already reserved is a potential race condition.  Adding the MAP_TRY_PLACED would only plug one side of that race condition, while doing nothing against the Vulkan mapping being overwritten.

Reviewer 2: Hrm... Maybe I've not woken up yet.  Yes, I think what @DadSchoorse suggested does work.  If `MAP_FIXED_NOREPLACE` succeeds, you've "reserved" the range and can replace it without worrying about stomping.  Ok, I think I"m convinced that the try isn't needed.

Reviewer 2: I've updated the spec to require replacement behavior.  This should get rid of any of that annoying "may have to reserve" nonsense.

the virtual address range via a platform-specific mechanism.
====

The sname:VkMemoryMapPlacedInfoEXT structure is defined as:

include::{generated}/api/structs/VkMemoryMapPlacedInfoEXT.txt[]

  * pname:sType is the type of this structure.
  * pname:pNext is `NULL` or a pointer to a structure extending this
    structure.
  * pname:pPlacedAddress is the virtual address at which to place the
    address.  If sname:VkMemoryMapInfoEXT::pname:flags does not contain
    ename:VK_MEMORY_MAP_PLACED_BIT_EXT, this value is ignored.

.Valid Usage
****
  * If sname:VkMemoryMapInfoEXT::pname:flags contains
    ename:VK_MEMORY_MAP_PLACED_BIT_EXT, pname:pPlacedAddress must: not be
    `NULL`
  * If sname:VkMemoryMapInfoEXT::pname:flags contains
    ename:VK_MEMORY_MAP_PLACED_BIT_EXT, pname:pPlacedAddress must: be
    aligned to an integer multiple of

>     sname:VkPhysicalDeviceMapMemoryPlacedPropertiesEXT::pname:minPlacedMemoryMapAlignment

Reviewer: Both of these VUs should move to the parent structure (VkMemoryMapInfoEXT) as they reference a value in the parent structure.

Reviewer 2: How about I drop the mention of flags and just say it has to be aligned.  0 is trivially aligned so it should be fine in 99% of the cases where the struct is unused.

Reviewer 2: I dropped the mention of flags in the alignment one.  The other is annoying.  It really is a VU on `pPlacedAddress` not the flags.  One option would be to toss the flag and just make it based on `VkMemoryMapPlacedInfoEXT` being present in the `pNext` chain and `pPlacedAddress != NULL`.  I kind-of do like the flag, though.

Reviewer: I'm fine with either resolution, though dropping the flag would make the API simpler, and make it more obvious that you can't do it with the vanilla map memory command.

    <<features-memoryMapPlaced, pname:memoryMapPlaced>> feature must: be
    enabled
  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags, the
    pname:pNext chain must: include a slink:VkMemoryMapPlacedInfoEXT
    structure
  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags and the
    <<features-memoryMapRangePlaced, pname:memoryMapRangePlaced>> feature
    is not enabled, pname:offset must: be zero
  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags and the
    <<features-memoryMapRangePlaced, pname:memoryMapRangePlaced>> feature
    is not enabled, pname:size must: be ename:VK_WHOLE_SIZE
  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags and the
    <<features-memoryMapRangePlaced, pname:memoryMapRangePlaced>> feature
    is enabled, pname:offset must: be aligned to an integer multiple of
    sname:VkPhysicalDeviceMapMemoryPlacedPropertiesEXT::pname:minPlacedMemoryMapAlignment
  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags and the
    <<features-memoryMapRangePlaced, pname:memoryMapRangePlaced>> feature
    is enabled, pname:size must: be ename:VK_WHOLE_SIZE or be aligned to an
    integer multiple of

>     sname:VkPhysicalDeviceMapMemoryPlacedPropertiesEXT::pname:minPlacedMemoryMapAlignment

Reviewer 6: VkDeviceMemory size is not required to be aligned and if caller would like to pass aligned size that includes the end of the buffer, it would break 'If size is not equal to VK_WHOLE_SIZE, size must be less than or equal to the size of the memory minus offset' requirement. I guess it's possible for the caller to use VK_WHOLE_SIZE in this case, but maybe it would make sense to remove this restriction, allow unaligned size from the caller and specify that implementation does the alignment?

Reviewer 6: Nevermind, I think I can see how it's better for drivers. It should be fine to handle that on Wine side.

include::{generated}/validity/protos/vkUnmapMemory.txt[]
--

ifdef::VK_EXT_map_memory2[]
[open,refpage='vkUnmapMemory2EXT',desc='Unmap a previously mapped memory object',type='protos']
--
Alternatively, to unmap a memory object once host access to it is no longer
needed by the application, call:

include::{generated}/api/protos/vkUnmapMemory2EXT.txt[]

  * pname:device is the logical device that owns the memory.
  * pname:pMemoryUnmapInfo is a pointer to a slink:VkMemoryUnmapInfoEXT
    structure describing parameters of the unmap.

This function behaves identically to flink:vkUnmapMemory except that it gets
its parameters via an extensible structure pointer rather than directly as
function arguments.

include::{generated}/validity/protos/vkUnmapMemory2EXT.txt[]
--

[open,refpage='VkMemoryUnmapInfoEXT',desc='Structure containing parameters of a memory unmap operation',type='structs']
--
The sname:VkMemoryUnmapInfoEXT structure is defined as:

include::{generated}/api/structs/VkMemoryUnmapInfoEXT.txt[]

  * pname:sType is the type of this structure.
  * pname:pNext is `NULL` or a pointer to a structure extending this
    structure.

>   * pname:memory is the slink:VkDeviceMemory object to be unmapped.

Reviewer: Missing the `flags` parameter?

Reviewer 2: Yup.  Fixed.

.Valid Usage
****

>   * pname:memory must: be currently host mapped

Reviewer: Just to check - this is expected to allow unmapping of memory mapped with VkMapMemory2EXT (and vice-versa), right?

Reviewer 2: Yes.  There's no reason why you can't mix+match.

Reviewer: Ok, don't think this needs calling out, just wanted to make sure it was expected.

Reviewer: (though this might need testing in the CTS test plan...)

ifdef::VK_EXT_map_memory_placed[]
  * If ename:VK_MEMORY_UNMAP_RESERVE_BIT_EXT is set in pname:flags,
    the <<features-memoryUnmapReserve, pname:memoryUnmapReserve>>
    must: be enabled
ifdef::VK_EXT_external_memory_host[]
  * If ename:VK_MEMORY_UNMAP_RESERVE_BIT_EXT is set in pname:flags, the
    memory object must: not have been imported from a handle type of
    ename:VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_ALLOCATION_BIT_EXT or
    ename:VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_MAPPED_FOREIGN_MEMORY_BIT_EXT
endif::VK_EXT_external_memory_host[]
endif::VK_EXT_map_memory_placed[]
****

include::{generated}/validity/structs/VkMemoryUnmapInfoEXT.txt[]
--

ifdef::VK_EXT_map_memory_placed[]
[open,refpage='VkMemoryUnmapFlagBitsEXT',desc='Bitmask specifying additional parameters of a memory unmap',type='enums']
--
Bits which can: be set in slink:VkMemoryUnmapInfoEXT::pname:flags, specifying
additional properties of a memory unmap, are:

include::{generated}/api/enums/VkMemoryUnmapFlagBitsEXT.txt[]

  * ename:VK_MEMORY_UNMAP_RESERVE_BIT_EXT requests that virtual address

>     range currently occupied by the memory map be remain reserved after the

Reviewer: ```suggestion
    range currently occupied by the memory map remains reserved after the
```

    flink:vkUnmapMemory2EXT call completes.
    Future system memory map operations or calls to flink:vkMapMemory or
    flink:vkMapMemory2EXT will not return addresses in that range unless

>     the range has since been unreserved by the client or the memory map

Reviewer: ```suggestion
    the range has since been unreserved by the client, the memory map
```

Reviewer 2: I think the original is correct.  It's a bit annoying the way it's phrased but I don't think the comma works.

Reviewer: Resolving this thread but only on the basis that the conversation is continuing in the thread below 🙃

>     operation is done with ename:VK_MEMORY_MAP_PLACED_BIT_EXT or the system

Reviewer: ```suggestion
    operation is done with ename:VK_MEMORY_MAP_PLACED_BIT_EXT, or the system
```

Reviewer 2: Same.  The last or is sort of a parenthetical.  It's supposed to read as unreserved || (VK_MAP_PLACED || SYS_MAP_FIXED)

Reviewer 3: I agree with @tobias - the comma disambiguates the grouping of the multiple 'or' clauses. It's somewhat like an 'eats shoots and leaves' exercise.

For complete disambiguation, though, a bullet list is probably better in this case.

Reviewer 2: Ok, I think I've improved it.  It now says:

> Future system memory map operations or calls to flink:vkMapMemory or
> `vkMapMemory2EXT` will not return addresses in that range unless
> the range has since been unreserved by the client or the mapping is
> explicitly placed in that range by calling `vkMapMemory2EXT` with
> `VK_MEMORY_MAP_PLACED_BIT_EXT` or doing the system memory map
> equivalent.

Reviewer 2: If we go with the explicit two bits I mentioned above, we could replace all this "reserve" nonsense and say that it replaces the Vulkan memory map with an inaccessible (`PROT_NONE`) anonymous map.

Reviewer: That's better, but I think the (or doing the system memory map equivalent) is something that should be in brackets. The lack of punctuation melts my brain 🙃

    memory map equivalent.
    When ename:VK_MEMORY_UNMAP_RESERVE_BIT_EXT is set, the memory unmap

>     operation may fail, in which case the memory object will remain host

Reviewer: ```suggestion
    operation may: fail, in which case the memory object will remain host
```

    mapped and flink:vkUnmapMemory2EXT will return

>     ename:VK_ERROR_MEMORY_MAP_FAILED.

Reviewer: Is it possible to quantify (perhaps in a note) the circumstances under which this happens?

Reviewer 2: Without digging into kernel source code, I'm not sure when it would fail.  I'm not even sure it can.  But `mmap()` is allowed to return `MAP_FAILED` for pretty arbitrary reasons and I would rather throw an error than silently fall back to `munmap()` and lose the reservation.

Reviewer 2: Context: https://gitlab.freedesktop.org/jekstrand/mesa/-/commit/348a18c3a559e280d8ef859fc6079ec44c1bde2e#637da2a63a937bb34d4a7112d341f9ea5c652765_168_177

Reviewer: Ok, if this is basically esoteric OS behavior then there's not much we can state here.
